### PR TITLE
NO-1592 Gas optimizations and code style recommendations

### PR DESCRIPTION
**Pull request by scottynomad**
- Switch to using _grantRole instead of _setup_role
- LNO-02C Remove redundant interface declaration
- SUS-01C store total cliffs
- SUS-02C loop limits
- SUS-03C remove use of signed int
- SUS-04C inefficient mapping lookup
- SUS-05C redundant lookup

### CODE REVIEW

**Comment by scottynomad** [comment]
Hmm, their gas optimization recommendations are a wash at best if not making things worse.

File: contracts/LockedNORI.sol [hunk 1]
```diff
@@ -402,18 +409,15 @@ contract LockedNORI is
...
+    _bridgedPolygonNori = bridgedPolygonNoriAddress;
     _ERC1820_REGISTRY.setInterfaceImplementer(
       address(this),
       ERC777_TOKENS_RECIPIENT_HASH,
       address(this)
     );
-    _setupRole(DEFAULT_ADMIN_ROLE, _msgSender()); // todo why doesnt grantRole work
-    _setupRole(TOKEN_GRANTER_ROLE, _msgSender()); // todo why doesnt grantRole work
-    _setupRole(PAUSER_ROLE, _msgSender()); // todo why doesnt grantRole work
+    _grantRole(TOKEN_GRANTER_ROLE, _msgSender());
```
Comment (jaycenhorton) [inline contracts/LockedNORI.sol:420]: is this intentionally excluding the PAUSER_ROLE?

File: contracts/ScheduleUtils.sol [hunk 2]
```diff
@@ -59,37 +60,25 @@ library ScheduleUtils {
...
+        time >= schedule.cliffs[cliffCount - 1].time,
         "Cliffs not chronological"
       );
     }
     require(time <= schedule.endTime, "Cliffs cannot end after schedule");
     require(
-      allCliffAmounts(schedule) + amount <= schedule.totalAmount,
-      "cliff amounts exceed total"
+      schedule.totalCliffAmount + amount <= schedule.totalAmount,
+      "Cliff amounts exceed total"
```
Comment (jaycenhorton) [inline contracts/ScheduleUtils.sol:75]:
```suggestion
      "ScheduleUtils: Cliff amounts exceed total"
```

File: contracts/ScheduleUtils.sol [hunk 3]
```diff
@@ -132,11 +122,10 @@ library ScheduleUtils {
       rampStartTime = schedule.cliffs[schedule.cliffCount - 1].time;
     }
     uint256 rampTotalTime = schedule.endTime - rampStartTime;
-    int256 rampTimeElapsed = int256(atTime) - int256(rampStartTime);
-    return
-      rampTimeElapsed <= 0
-        ? 0
-        : (rampTotalAmount * uint256(rampTimeElapsed)) / rampTotalTime;
+    if (atTime < rampStartTime) {
```
Comment (jaycenhorton) [inline contracts/ScheduleUtils.sol:125]: was removing the ternary recommended in favor of multi-return?

File: contracts/LockedNORI.sol (hunk 1 above)
Comment (scottynomad) [reply contracts/LockedNORI.sol:420]: I introduce a call to `__ERC777PresetPausablePermissioned_init_unchained();` just above which sets up PAUSER and ADMIN roles consistently with how we do it in the bridged nori contract.

File: contracts/ScheduleUtils.sol (hunk 3 above)
Comment (scottynomad) [reply contracts/ScheduleUtils.sol:125]: not explicitly no, it was just about not using signed ints.  I'll factor tat back out.

**Review by jaycenhorton — APPROVED** [review]

**Pull request merged**